### PR TITLE
Introducing V1 grpc control protocol for backward compatibility with older agents

### PR DESCRIPTION
**Pull request by michalpristas**
## What does this PR do?

Introducing V1 grpc control protocol for backward compatibility with older agents trying to upgrade.
Only necessary methods implemented, calling pprof or metrics will result in UNIMPLEMENTED result

Conflicting fields are renamed to include V1 Prefix.
Conflict happens in between 2 `proto` namespaces one is here on V1 control protocol, the other one is in `elastic-agent-client` definition. 
As differentiating them by namespace would be complicated in a sense of handling adoption in such a tense deadline renaming seems like a feasible choice and is not breaking with serialization we use. 

Tested upgrade from 8.5.3 to locally build package. No rollback.

## Why is it important?

Without having backwards compatible protocol, watcher initiated from 8.5.3 fails to fetch agent status and evaluates this as failed upgrade. This leads to rollback and 8.6 is not installed after all.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

Fixes: #1924

### CODE REVIEW

File: control_v1.proto [hunk 1]
```diff
@@ -0,0 +1,119 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+syntax = "proto3";
+
+package proto;
```
Comment (cmacknz) [inline control_v1.proto:7]: Let's add a comment that this is the same package as https://github.com/elastic/elastic-agent-client/blob/15881a8e64ef95f32f688c933319cd926ff500b0/elastic-agent-client.proto#L7 and we need to avoid name collisions with it.

Not that we should be editing this file after this.

File: internal/pkg/agent/control/v1/client/client.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	UserID             string
+	UserGID            string
+	BinaryArchitecture string
+	RouteKey           string
+	ElasticLicensed    bool
+	Error              string
+}
+
+// ProcPProf returns pprof data for a process.
+type ProcPProf struct {
```
Comment (cmacknz) [inline internal/pkg/agent/control/v1/client/client.go:76]: I think this and ProcMeta above are unused.

File: internal/pkg/agent/control/v1/server/server.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+package server
+
+import (
+	"context"
+
+	"go.elastic.co/apm"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/control"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/control/v1/proto"
+	cproto "github.com/elastic/elastic-agent/internal/pkg/agent/control/v2/cproto"
```
Comment (cmacknz) [inline internal/pkg/agent/control/v1/server/server.go:14]:
```suggestion
	v2proto "github.com/elastic/elastic-agent/internal/pkg/agent/control/v2/cproto"
```

File: internal/pkg/agent/control/v2/server/server.go [hunk 4]
```diff
@@ -77,6 +79,11 @@ func (s *Server) Start() error {
 	}
 	cproto.RegisterElasticAgentControlServer(s.server, s)
 
+	if startV1Wrapper {
```
Comment (cmacknz) [inline internal/pkg/agent/control/v2/server/server.go:82]: Why do we need to make this conditional?

Also can you add a comment or two explaining why we need the V1 wrapper so we can remember it in 6 months?

File: internal/pkg/agent/control/v1/server/server.go (hunk 3 above)
Comment (cmacknz) [reply internal/pkg/agent/control/v1/server/server.go:14]: This makes it easier to spot which version is used where.

**Review by cmacknz — APPROVED** [review]
I think this could use some more explanation of why it exists directly in the code, but none of my comments are blocking.

Thanks!

File: control_v2.proto [hunk 5]
```diff
@@ -106,7 +106,7 @@ message ComponentUnitState {
   UnitType unit_type = 1;
   // ID of the unit in the component.
   string unit_id = 2;
-  // Current state.
+  // Current status.
```
Comment (blakerouse) [inline control_v2.proto:109]: Why the change to status here? Probably no benefit to changing the contents of this file, minus the `go_package` above.

File: internal/pkg/agent/control/v1/server/server.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	if state == v2proto.State_FAILED {
+		return proto.Status_V1_FAILED
+	}
+	return proto.Status_V1_HEALTHY
+}
+
+func componentStateToProto(components []*v2proto.ComponentState) []*proto.ApplicationStatus {
+	s := make([]*proto.ApplicationStatus, len(components))
+	for i, c := range components {
+		s[i] = &proto.ApplicationStatus{
```
Comment (blakerouse) [inline internal/pkg/agent/control/v1/server/server.go:107]: In V2 each unit has its own state separate from the component. So it's possible that a v2 component be healthy and the unit inside of it be not healthy. Should we take the worse state of the unit and propagate it as the status of the application?

**Review by blakerouse — APPROVED** [review]
Looks really good, had 2 comments.

File: control_v2.proto (hunk 5 above)
Comment (michalpristas) [reply control_v2.proto:109]: thanks for noticing this is just a leftover from refactorings

File: internal/pkg/agent/control/v1/server/server.go (hunk 6 above)
Comment (cmacknz) [reply internal/pkg/agent/control/v1/server/server.go:118]: If we do this, any integration that is unhealthy can cause a rollback. This may lead to unnecessary rollbacks.

We also have no idea if we are rolling back to an agent that is any better behaved. 

The most conservative approach is to leave this as is and not rollback as long as the components themselves have started and are healthy.

Comment (cmacknz) [reply internal/pkg/agent/control/v1/server/server.go:118]: Probably considering the unit state better maps to what V1 did though.
